### DISCUSSION TITLE
Adding details via code

### DISCUSSION THREAD

**Discussion by Skosn** [visualdesigncafe/nature-renderer#13] [Ideas]
Hello,

if I'm correct as of now you can only register details with LODs when adding them directly to the NatureRenderer component via editor.
It would be nice to be able to also add them via API during runtime.
Or do LODs get fetched from the terrain prototypes during runtime? If so the docs state otherwise.

**Comment by maxplooi** [comment]
You can use the regular API of Unity to register details at runtime, it is fully supported. 

The main reason you need to use the component in the editor instead of the terrain inspector is because the terrain inspector automatically blocks prefabs with LODs. You'll see a _LOD Group component is not supported_ error in the UI. This only happens in the UI, not when using the API.

Note that Nature Renderer automatically replaces prefabs with LODs with a placeholder (usually LOD0 of the prefab, instead of the root). This is a workaround for some internal issues in Unity. It happens automatically when registering details using the API so there is no need to worry about it. However, it causes one additional terrain refresh at startup, so if performance is very important then you can use the following API instead:

**To register a prefab:**
```c#
using VisualDesignCafe.Rendering.Nature

var terrain = GetComponent<Terrain>();
terrain.terrainData.detailPrototypes = 
    new[] 
    {
        new DetailPrototype()
        {
            // Instead of "prototype = YOUR_PREFAB"
            prototype = NatureRendererUtility.GetSupportedPlaceholder( YOUR_PREFAB );
        }
    }
``` 

**To get a prefab:**
```c#
using VisualDesignCafe.Rendering.Nature

var terrain = GetComponent<Terrain>();
var prototypes = terrain.terrainData.detailPrototypes;

// Instead of "prefab = prototypes[0].prototype"
GameObject prefab = NatureRendererUtility.GetPrototypeToRender( prototypes[0] );
```
